src/pcre2_jit_compile.c

```diff
@@ -12979,6 +12984,7 @@ while (current)
 
     case OP_BRAMINZERO:
     compile_braminzero_backtrackingpath(common, current);
+    // XXX what if error?
```

Comment:
It should not cause any error. An idea:
https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitLir.h#L688

`sljit_set_compiler_memory_error()` can be used to kill code compilation. Another (global) flag to stop malloc. Then we can see what happens.

Reply:
I see. JIT allocates memory from its own small slab, rather than through PCRE2's malloc callback. So, the changes in pcre2test aren't exercising those allocation failures?

Reply:
Note: `SLJIT_MALLOC` uses the mem controller malloc.
https://github.com/PCRE2Project/pcre2/blob/master/src/pcre2_jit_compile.c#L66
https://github.com/PCRE2Project/pcre2/blob/master/src/pcre2_jit_compile.c#L72

Reply:
Could we add a flag or mode so that all allocations go through malloc (new "fragment") rather than reusing a previous one? Then we'd be able to exercise all the failures more easily. Would that be best with a runtime flag, or a compile flag? It looks like we'd need to make a change in the sljit repo, not just in pcre2.

Reply:
I'm also a bit worried by the two `return NULL` cases in `sljit_alloc_memory`. They don't set an error! So _if_ they were hit (and they shouldn't be...) then the calling code wouldn't necessarily notice. We probably have sufficient testing however, to ensure that callers aren't requesting more than 64/128 bytes.

Reply:
> `PTR_FAIL_IF_NULL` sets the error code.

I think the case of size > 64/128 bytes doesn't do that, and just returns directly without setting an error condition?

Reply:
That is an api misuse. An assert could be useful.

Reply:
I tried making a change to sjlitLir ensure_buf/ensure_abuf. I made it so that they always allocated a fresh fragment.

Unfortunately... they allocate a _lot_ of fragments, and it takes a looooong time for the tests to successively test the behaviour when each one fails.

For the time being, I'm just going to leave this something which isn't tested automatically.

We can be positive though: the error-handling in scheme in the JIT code is very good, and the compiler really does handle malloc failures very well, as far as I can tell.